### ISSUE TITLE
Allow User to Queue Files Before Upload (a.k.a. manual upload start)

### ISSUE THREAD

**Issue by noamichael** [vaadin/vaadin-upload#123]
After filling out an HTML form, it is generally submitted to the sever in one POST request after it's been validated. This is impossible if the form uses the `<vaadin-upload/>` component because it eagerly uploaded files right after the user makes their selection from the drive. A nice feature would be to allow for an attribute such as `postponeUpload` which would allow page author to manually upload the files when ready or be able to send additional fields with the request.

Example:

```
<template>
    <input type="text" id="fileDescription" required placeholder="Enter File Description">
    <vaadin-upload id="fileUploader" postpone-upload target="api/v1/upload"></vaadin-upload>
    <button on-click="onSubmit">Submit</button>
</template>
<script>
    Polymer({
           is: 'some-component',
           onSubmit: function(){
                 let fileDescription = this.$.fileDescription.value;
                  if(!fileDescription){return;}
                  this.$.fileUploader.uploadFiles({
                             fileDescription: fileDescription//pass additional form fields for request
                  });
           }
    });
</script>
```

**Comment by jouni** [comment]
HI, and thanks for the feedback! This feature is definitely high on our priority list, but we’re currently focusing on other elements in our collection.

If you want to prototype an implementation for this, and send a PR, we’d be happy to help you with it.

**Comment by noamichael** [comment]
@jouni I'll try to write a prototype this weekend which will function much like the example I made above. My initial instinct is to still let the component do the POST/PUT of the file and let the component user attach any additional formData items.  Maybe the attribute could be named `manual-upload-start` instead of `postpone-upload`.

**Comment by jay8t6** [comment]
+1 

I was just looking to see if this feature was already enabled. I think this is pretty important if we were to use this element on forms. It works great as a standalone element.

**Comment by noamichael** [comment]
@jay8t6 @jouni I'm look for some opinions on how the component should function in a form. I'm imagining that on submit, all the files queued would be uploaded all at once.

1. How should the progress be displayed? Currently, it's all individual progress bars. But if all files are uploaded at once, should there be only one progress bar?

**Comment by noamichael** [comment]
@jay8t6 @jouni I'm looking for some opinions on how the component should function in a form. I'm imagining that on submit, all the files queued would be uploaded at once. I'm trying to decide if all files should be in a single  `XMLHttpRequest`, or individually like they are now. This may have to be a setting because I can't predict what the users REST API might look like, but a single request would require a major rethinking of the current implementation.  Here's an example of the API I have in mind if each file gets it's own request:

    <template>
        <input type="text" id="fileDescription" required placeholder="Enter File Description">
        <vaadin-upload id="fileUploader" manual-upload-start target="api/v1/upload"></vaadin-upload>
        <button on-click="onSubmit">Submit</button>
    </template>
    <script>
        Polymer({
               is: 'some-component',
               onSubmit: function(){
                     //user performs some custom validation
                     let fileDescription = this.$.fileDescription.value;
                     if(!fileDescription){return;}

                     //User calls public facing API method. 
                     this.$.fileUploader.uploadFiles({
                          formData: { //data to append to request
                              fileDescription: fileDescription
                          },
                          fileUploaded: function(response){ 
                               //callback for each file's success
                          },
                          complete: function(response){
                               //all files uploaded
                          },
                          error: function(file, response){
                               //callback when a file fails to upload
                          },
                          progress: function(file, percent){ 
                               //callback for each files progress
                          }
                      });
               }
        });
    </script>

What are your thoughts?

**Comment by jouni** [comment]
@noamichael, I can’t say much about how a dev would want to handle the XHRs, would it be better to have all files in one. So for me, keeping the current separate ones is preferred, but I can’t really justify this opinion with anything :)

As for the end user, it also seems like there’s no need to “combine” the uploads. Having a separate progress indicator for each file should be fine, I think.

**Comment by jouni** [comment]
@platosha and @manolo might have some other opinions or ideas.

**Comment by platosha** [comment]
Implemented in #168
